pkg/benchmark/pod_container.go

```diff
@@ -69,18 +66,31 @@ var _ = framework.KubeDescribe("PodSandbox", func() {
 				Linux:    &runtimeapi.LinuxPodSandboxConfig{},
 			}
 
-			operation := b.Time("create PodSandbox and container", func() {
+			benchmark := func() {
 				By("run PodSandbox")
-				podID, err = rc.RunPodSandbox(config, framework.TestContext.RuntimeHandler)
+				podID, err := rc.RunPodSandbox(config, framework.TestContext.RuntimeHandler)
 				framework.ExpectNoError(err, "failed to create PodSandbox: %v", err)
+
 				By("create container in PodSandbox")
 				containerID := framework.CreateDefaultContainer(rc, ic, podID, config, "Pod-Container-for-creating-benchmark-")
+
 				By("start container in PodSandbox")
 				err = rc.StartContainer(containerID)
-			})
+				framework.ExpectNoError(err, "failed to start Container: %v", err)
+
+				By("stop PodSandbox")
+				rc.StopPodSandbox(podID)
+				By("delete PodSandbox")
+				rc.RemovePodSandbox(podID)
+			}
+
+			// Run a single test to ensure images are available and everything works
+			benchmark()
+
+			// Do the benchmark
```

Comment:
immediately after running a benchmark the runtimes are likely to be busy removing image snapshots, hitting golang gc, and such... running the benchmark twice in a row is a different benchmark.

Reply:
May I ask you to clarify for me what we should do here? We pre-run the benchmark once to ensure everything is cached and in place (images for example).

Reply:
@mikebrow do you mean we should do this in a different way (e.g. invoking image pull API) to avoid delete operations here?

Reply:
Just pointing out that gc of image and golang objects/artifacts, are async tasks that lazily follow removing the references (count of containers using an image and snapshots, references in memory to golang memory objects... that lazy activity is likely to run during the 2nd run if the 2nd run immediately follows. Said gc activities will probably use locks that will delay the 2nd run. You could wait between runs for the system to settle down.. or measure both runs as users rarely start a dummy pod to get the system ready to run a second pod.

if you turn on debug you'll see the gc outputs

if you want to pre pull the images, and they have a specific tag (not latest, and not untagged which defaults to latest) then you can use no-pull as that will eliminate the timing issue related to remote image pull from the registry. However then you are not benchmarking the image pull, possibly not the unpack (I'd have to check) :-) (image pull, unpack, and snapshot for a particular container before adding it's volumes for the roots are the expensive image resource operations)  

benchmarking is hard...